web-client/public/js/graph.js

```diff
@@ -112,46 +124,71 @@ var drawGraph = function (nodes, links, positiveWeights, negativeWeights, sheetT
                      .append("g")
 
 
+  var previousValue = zoom.translate().slice();
+  var previousMax = [width, height];
+  var previousScale = zoom.scale();
   function zoomed(manual = false) {
+    var scaleRange = maximumScale - minimumScale;
+    var percentZoom = zoom.scale() / scaleRange * 100;
     if (!MANUAL_ZOOM) {
       $(".zoomSlider").val(d3.event.scale.toFixed(2)); // This doesn't work using d3 selection for some reason
     }
-    svg.attr("transform", "translate(" + d3.event.translate + ")scale(" + d3.event.scale + ")");
-  }
-
-  d3.selectAll(".scrollBtn").on("click", null); // Remove event handlers, if there were any.
-
-  // TODO: Make this less bad
-  d3.select(".scrollUp").on("click", function () {
-    move("up");
-  });
+    if (!adaptive) { // Limit to viewport
+      var scale = zoom.scale();
+      var scaledWidth = scale * width;
+      var scaledHeight = scale * height;
+      var maxX = width - scaledWidth;
+      var maxY = height - scaledHeight;
+
+      if (d3.event.translate[0] < 0) {
+        d3.event.translate[0] = 0;
+      } else if (d3.event.translate[0] > maxX) {
+        d3.event.translate[0] = maxX;
+      }
 
-  d3.select(".scrollLeft").on("click", function () {
-    move("left");
-  });
+      if (d3.event.translate[1] < 0) {
+        d3.event.translate[1] = 0;
+      } else if (d3.event.translate[1] > maxY) {
+        d3.event.translate[1] = maxY;
+      }
 
-  d3.select(".scrollRight").on("click", function () {
-    move("right");
-  });
+      zoom.translate([d3.event.translate[0], d3.event.translate[1]]);
+    }
+    if (!scrolling && d3.event.scale < 1) {
+      $container.addClass("cursorGrab");
+      scrolling = true;
+    } else if (!adaptive && scrolling && d3.event.scale >= 1) {
+      if ($container.hasClass("cursorGrab")) $container.removeClass("cursorGrab");
+      if ($container.hasClass("cursorGrabbing")) $container.removeClass("cursorGrabbing");
+      scrolling = false;
+    }
+    // $("#zoomPercent").html(percentZoom.toFixed(1) + "%");
```

Comment:
The commented out lines all mentioned here are for the implementations of #465 that haven't been put in yet. I'll remove them in the PR and add them back later.